### PR TITLE
Avoid reconnect in Load Balancers

### PR DESCRIPTION
**Pull request by ejona86**
You can look at the commits separately, or together. Still missing tests.

CC @zhangkun83

### CODE REVIEW

File: core/src/main/java/io/grpc/internal/InternalSubchannel.java [hunk 1]
```diff
@@ -284,6 +287,42 @@ public void run() {
...
+          SocketAddress address = oldAddressGroup.getAddresses().get(addressIndex);
+          int newIndex = newAddressGroup.getAddresses().indexOf(address);
+          if (newIndex != -1) {
+            addressIndex = newIndex;
+          } else {
+            // Forced to drop the connection
+            if (state.getState() == READY) {
+              savedTransport = activeTransport;
+              activeTransport = null;
+              addressIndex = 0;
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/InternalSubchannel.java]: nit: move this line outside of `if`?

Comment (ejona86) [inline core/src/main/java/io/grpc/internal/InternalSubchannel.java]: meh. I like it being adjacent the `activeTransport`/`pendingTransport = null`. I think the added verbosity makes it more readable.

File: core/src/test/java/io/grpc/internal/InternalSubchannelTest.java [hunk 2]
```diff
@@ -381,6 +381,199 @@ protected void onNotInUse(InternalSubchannel is) {
...
+    assertEquals(ConnectivityState.CONNECTING, internalSubchannel.getState());
+
+    // Second address connecting
+    verify(mockTransportFactory).newClientTransport(addr2, AUTHORITY, USER_AGENT);
+    assertNoCallbackInvoke();
+    assertEquals(ConnectivityState.CONNECTING, internalSubchannel.getState());
+
+    // Update addresses
+    internalSubchannel.updateAddresses(new EquivalentAddressGroup(Arrays.asList(addr3, addr4)));
+    assertNoCallbackInvoke();
```
Comment (ejona86) [inline core/src/test/java/io/grpc/internal/InternalSubchannelTest.java]: I verify it already on line 562.

File: core/src/test/java/io/grpc/internal/InternalSubchannelTest.java [hunk 3]
```diff
@@ -381,6 +381,199 @@ protected void onNotInUse(InternalSubchannel is) {
...
+
+    // Second address connects
+    verify(mockTransportFactory).newClientTransport(addr2, AUTHORITY, USER_AGENT);
+    transports.peek().listener.transportReady();
+    assertExactCallbackInvokes("onStateChange:READY");
+    assertEquals(ConnectivityState.READY, internalSubchannel.getState());
+
+    // Update addresses
+    internalSubchannel.updateAddresses(new EquivalentAddressGroup(Arrays.asList(addr3, addr4)));
+    assertExactCallbackInvokes("onStateChange:IDLE");
```
Comment (ejona86) [inline core/src/test/java/io/grpc/internal/InternalSubchannelTest.java]: Done.

File: core/src/test/java/io/grpc/internal/InternalSubchannelTest.java [hunk 4]
```diff
@@ -381,6 +381,199 @@ protected void onNotInUse(InternalSubchannel is) {
...
+    assertEquals(ConnectivityState.CONNECTING, internalSubchannel.getState());
+
+    // Second address connecting
+    verify(mockTransportFactory).newClientTransport(addr2, AUTHORITY, USER_AGENT);
+    assertNoCallbackInvoke();
+    assertEquals(ConnectivityState.CONNECTING, internalSubchannel.getState());
+
+    // Update addresses
+    internalSubchannel.updateAddresses(new EquivalentAddressGroup(Arrays.asList(addr2, addr3)));
+    assertNoCallbackInvoke();
```
Comment (ejona86) [inline core/src/test/java/io/grpc/internal/InternalSubchannelTest.java]: Done.

File: core/src/test/java/io/grpc/internal/InternalSubchannelTest.java [hunk 5]
```diff
@@ -381,6 +381,199 @@ protected void onNotInUse(InternalSubchannel is) {
...
+
+    // Second address connects
+    verify(mockTransportFactory).newClientTransport(addr2, AUTHORITY, USER_AGENT);
+    transports.peek().listener.transportReady();
+    assertExactCallbackInvokes("onStateChange:READY");
+    assertEquals(ConnectivityState.READY, internalSubchannel.getState());
+
+    // Update addresses
+    internalSubchannel.updateAddresses(new EquivalentAddressGroup(Arrays.asList(addr2, addr3)));
+    assertNoCallbackInvoke();
```
Comment (ejona86) [inline core/src/test/java/io/grpc/internal/InternalSubchannelTest.java]: Done.

**Pull request merged**